File: androidWeeFood/src/main/java/de/darthkali/weefood/android/presentation/screens/new_recipe/NewRecipeViewModel.kt
package de.darthkali.weefood.android.presentation.screens.new_recipe

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import de.darthkali.weefood.android.presentation.screens.BaseViewModel
import de.darthkali.weefood.domain.model.Recipe
import de.darthkali.weefood.interactors.recipe.GetRecipe
import de.darthkali.weefood.interactors.recipe.SaveRecipe
import de.darthkali.weefood.interactors.recipe_ingredient.DeleteRecipeIngredient
import de.darthkali.weefood.interactors.recipe_ingredient.GetIngredientsFromRecipe
import de.darthkali.weefood.presentation.new_recipe.NewRecipeEvents
import de.darthkali.weefood.presentation.new_recipe.NewRecipeState
import de.darthkali.weefood.util.Logger
import org.koin.core.component.inject

class NewRecipeViewModel(
    recipeId: Int,
) : BaseViewModel() {

    private val logger = Logger("NewRecipeViewModel")
    private val saveRecipe: SaveRecipe by inject()
    private val getRecipe: GetRecipe by inject()
    private val deleteRecipeIngredient: DeleteRecipeIngredient by inject()
    private val getIngredientsFromRecipe: GetIngredientsFromRecipe by inject()

    val state: MutableState<NewRecipeState> = mutableStateOf(NewRecipeState())

    init {
        if (recipeId != 0) { //TODO: vlt nicht nötig, da wir im getRecipe schon auf > 0 prüfen
            onTriggerEvent(NewRecipeEvents.GetRecipe(recipeId = recipeId))
        }
    }


    fun onTriggerEvent(event: NewRecipeEvents) {
        when (event) {
            is NewRecipeEvents.GetRecipe -> {

                val recipe: Recipe = getRecipe.execute(recipeId = event.recipeId)!!
                state.value = state.value.copy(recipe = recipe)

            }
            is NewRecipeEvents.OnUpdateName -> {
                onUpdateRecipe(state.value.recipe.copy(name = event.name))
            }
            is NewRecipeEvents.OnUpdateImage -> {
                onUpdateRecipe(state.value.recipe.copy(image = event.image))
            }
            is NewRecipeEvents.OnUpdateCookingTime -> {
                onUpdateRecipe(state.value.recipe.copy(cooking_time = event.cooking_time))
            }
            is NewRecipeEvents.OnUpdateCookingTimeUnit -> {
                onUpdateRecipe(state.value.recipe.copy(cooking_time_unit = event.cooking_time_unit))
            }
            is NewRecipeEvents.OnUpdateDescription -> {
                onUpdateRecipe(state.value.recipe.copy(recipeDescription = event.description))
            }
            is NewRecipeEvents.OnAddIngredient -> {
                onUpdateRecipe(state.value.recipe.copy(databaseId = saveRecipe.execute(state.value.recipe)))
            }
            is NewRecipeEvents.OnDeleteIngredient -> {
                if (deleteRecipeIngredient.execute(
                        state.value.recipe.databaseId!!,
                        event.ingredient.internalId!!
                    )
                ) {
                    val ingredients =
                        getIngredientsFromRecipe.execute(state.value.recipe.databaseId!!)
                    onUpdateRecipe(state.value.recipe.copy(ingredients = ingredients!!))
                }
            }
            is NewRecipeEvents.OnSaveRecipe -> {
                saveRecipe.execute(state.value.recipe)
            }
            else -> {
                logger.log("Something went wrong.")
            }
        }
    }

    private fun onUpdateRecipe(recipe: Recipe) {
        state.value = state.value.copy(recipe = recipe, changed = state.value.changed + 1)
    }

}

File: iosWeeFood/iosWeeFood/presentation/screens/new_recipe/NewRecipeViewModel.swift
//
//  NewRecipeViewModel.swift
//  iosWeeFood
//
//  Created by Danny Steinbrecher on 10.08.21.
//
import SwiftUI
import shared

class NewRecipeViewModel: ObservableObject {
    
    private let logger = Logger(className: "NewRecipeViewModel")

    // Dependencies
    private let saveRecipe =  SaveRecipe()
    private let getRecipe = GetRecipe()
    private let deleteRecipeIngredient = DeleteRecipeIngredient()
    private let getIngredientsFromRecipe = GetIngredientsFromRecipe()
   

    // State
    @Published var state: RecipeDetailState =  RecipeDetailState()

    init(
        recipeId: Int
    ) {
        onTriggerEvent(event: RecipeDetailEvents.GetRecipe(recipeId: Int32(recipeId)))
    }

    func onTriggerEvent(event: RecipeDetailEvents){
        var currentRecipe = (self.state.recipe.copy() as! Recipe)
        
        switch event {
            case is RecipeDetailEvents.GetRecipe:
                
                let recipeId = Int32((event as! RecipeDetailEvents.GetRecipe).recipeId)
                
                let emptyRecipe = Recipe(
                    databaseId: 0,
                    name: "",
                    image: "",
                    cooking_time: 0,
                    cooking_time_unit: "",
                    recipeDescription: "",
                    portion: 0,
                    ingredients: [])
                currentRecipe = getRecipe.execute(recipeId: recipeId) ?? emptyRecipe
                
                
                self.updateState(recipe: currentRecipe)
                
            case is RecipeDetailEvents.OnUpdateName:
                currentRecipe.name = (event as! RecipeDetailEvents.OnUpdateName).name
                self.updateState(recipe: currentRecipe)

            case is RecipeDetailEvents.OnUpdateImage:
                currentRecipe.image = (event as! RecipeDetailEvents.OnUpdateImage).image
                self.updateState(recipe: currentRecipe)

            case is RecipeDetailEvents.OnUpdateCookingTime:
                currentRecipe.cooking_time = Int32(exactly: (event as! RecipeDetailEvents.OnUpdateCookingTime).cooking_time) as? KotlinInt
                self.updateState(recipe: currentRecipe)
                
            case is RecipeDetailEvents.OnUpdateCookingTimeUnit:
                currentRecipe.cooking_time_unit = (event as! RecipeDetailEvents.OnUpdateCookingTimeUnit).cooking_time_unit
                self.updateState(recipe: currentRecipe)
                
            case is RecipeDetailEvents.OnUpdateDescription:
                currentRecipe.recipeDescription = (event as! RecipeDetailEvents.OnUpdateDescription).description
                self.updateState(recipe: currentRecipe)
                
            case is RecipeDetailEvents.OnAddIngredient:
                
                currentRecipe.databaseId = saveRecipe.execute(recipe: (event as! RecipeDetailEvents.OnAddIngredient).recipe)
                self.updateState(recipe: currentRecipe)
                
          /* case is NewRecipeEvents.OnDeleteIngredient:
                if (deleteRecipeIngredient.execute(
                        state.value.recipe.databaseId!!,
                        event.ingredient.internalId!!
                    )
                ) {
                    val ingredients =
                        getIngredientsFromRecipe.execute(state.value.recipe.databaseId!!)
                    self.updateState(state.value.recipe.copy(ingredients = ingredients!!))
                }*/
            case is RecipeDetailEvents.OnSaveRecipe:
                currentRecipe.databaseId = saveRecipe.execute(recipe: state.recipe)
                self.updateState(recipe: currentRecipe)

            default: doNothing()
        }
    }



    private func doNothing(){}

    private func updateState(
        recipe: Recipe? = nil
    ){
        let currentState = (self.state.copy() as! RecipeDetailState)
        self.state = self.state.doCopy(
            changed: currentState.changed + 1,
            recipe:recipe ?? currentState.recipe
        )
    }
    


}
